src/lib/optimizer/strategy/stored_table_column_alignment_rule.hpp

```diff
@@ -0,0 +1,54 @@
+#pragma once
+
+#include <unordered_map>
+#include <vector>
+
+#include "abstract_rule.hpp"
+#include "logical_query_plan/stored_table_node.hpp"
+
+namespace opossum {
+
+// Modified hash code generation for StoredTableNodes where column pruning information is omitted. Struct is used to
+// enable hash based containers containing std::shared_ptr<StoredTableNode>.
+struct StoredTableNodeSharedPtrHash final {
+  size_t operator()(const std::shared_ptr<StoredTableNode>& node) const {
+    size_t hash{0};
+    boost::hash_combine(hash, node->table_name);
+    for (const auto& pruned_chunk_id : node->pruned_chunk_ids()) {
+      boost::hash_combine(hash, static_cast<size_t>(pruned_chunk_id));
+    }
+    return hash;
+  }
+};
+
+// Modified equals evaluation code for StoredTableNodes where column pruning information is omitted. Struct is used to
+// enable hash based containers containing std::shared_ptr<StoredTableNode>.
+struct StoredTableNodeSharedPtrEqual final {
+  size_t operator()(const std::shared_ptr<StoredTableNode>& lhs, const std::shared_ptr<StoredTableNode>& rhs) const {
+    return lhs == rhs || (lhs->table_name == rhs->table_name && lhs->pruned_chunk_ids() == rhs->pruned_chunk_ids());
+  }
+};
+
+template <typename Value>
+using StoredTableNodeUnorderedMap = std::unordered_map<std::shared_ptr<StoredTableNode>, Value,
+                                                       StoredTableNodeSharedPtrHash, StoredTableNodeSharedPtrEqual>;
+
+// The PQP sub plan memoization for StoredTableNodes is sensitive to the node's table name, set of pruned chunks and
+// set of pruned columns. Consequently, for multiple nodes with the same table name, chunk and column pruning, only one
+// GetTable operator is created and executed. In some queries, the ColumnPruningRule and ChunkPruningRule provide an
+// LQP with multiple StoredTableNodes where the table names and sets of pruned chunks are equal but the sets of pruned
+// columns are different. This leads to the creation and execution of different GetTable operators.
```

Comment:
I'm not sure if the comment is helpful for someone who has not seen the "broken" LQP. I wonder if an example would be helpful.

Reply:
"creation and execution of different GetTable operators." > leads to GetTables not being merged?

Also, does this rule only apply to pruned columns or also pruned chunks? The Q21 case would probably still be worth merging the nodes even if one GetTable node has "only a few" chunks more to load. Or am I completely on the wrong path here?

Reply:
My gut feeling is that the case of different chunk pruning lists is much more unlikely. The problem tackled in this PR is that we have two operators operating on the same data, but the translator is unable to identify it. If the chunk pruning lists differ, we are not working on the same data any more.

Not sure if this explanation makes perfect sense. In any case, I would only look at it once we actually see it in PQPs.

Reply:
Ok

Reply:
> I'm not sure if the comment is helpful for someone who has not seen the "broken" LQP. I wonder if an example would be helpful.

We have an example now. But maybe the .hpp is a little long now. Alternatively, if you think it's too long now, we could put the example on a wiki page and reference it in the .hpp file?

Reply:
I think comments are more likely to be read and updated than wiki pages. Having to scroll just a bit more is a small price to pay.